addons/gamification/models/gamification_karma_tracking.py

```diff
@@ -1,69 +1,133 @@
-        """ Consolidate last month. Used by a cron to cleanup tracking records. """
-        previous_month_start = fields.Date.today() + relativedelta(months=-1, day=1)
-        return self._process_consolidate(previous_month_start)
-
-    def _process_consolidate(self, from_date):
-        """ Consolidate trackings into a single record for a given month, starting
-        at a from_date (included). End date is set to last day of current month
-        using a smart calendar.monthrange construction. """
-        end_date = from_date + relativedelta(day=calendar.monthrange(from_date.year, from_date.month)[1])
+    def _consolidate_cron(self):
+        """Consolidate the trackings 2 months ago. Used by a cron to cleanup tracking records."""
+        from_date = date_utils.start_of(fields.Datetime.today(), 'month') - relativedelta(months=2)
+        return self._process_consolidate(from_date)
+
+    def _process_consolidate(self, from_date, end_date=None):
+        """Consolidate the karma trackings.
+
+        The consolidation keeps, for each user, the oldest "old_value" and the most recent
+        "new_value", creates a new karma tracking with those values and removes all karma
+        trackings between those dates.
+        """
+        self.env['gamification.karma.tracking'].flush_model()
+
+        if not end_date:
+            end_date = date_utils.end_of(date_utils.end_of(from_date, 'month'), 'day')
+
         select_query = """
-SELECT user_id,
-(
-    SELECT old_value from gamification_karma_tracking old_tracking
-    WHERE old_tracking.user_id = gamification_karma_tracking.user_id
-        AND tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-        AND consolidated IS NOT TRUE
-        ORDER BY tracking_date ASC LIMIT 1
-), (
-    SELECT new_value from gamification_karma_tracking new_tracking
-    WHERE new_tracking.user_id = gamification_karma_tracking.user_id
-        AND tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-        AND consolidated IS NOT TRUE
-        ORDER BY tracking_date DESC LIMIT 1
-)
-FROM gamification_karma_tracking
-WHERE tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-AND consolidated IS NOT TRUE
-GROUP BY user_id """
+        WITH old_tracking AS (
+            SELECT DISTINCT ON (user_id) user_id, old_value, tracking_date
+              FROM gamification_karma_tracking
+             WHERE tracking_date BETWEEN %(from_date)s
+               AND %(end_date)s
+               AND consolidated IS NOT TRUE
+          ORDER BY user_id, tracking_date ASC, id ASC
+        )
+            INSERT INTO gamification_karma_tracking (
+                            user_id,
+                            old_value,
+                            new_value,
+                            tracking_date,
+                            origin_ref,
+                            consolidated,
+                            reason)
+            SELECT DISTINCT ON (nt.user_id)
```

Comment:
Could be great to tell in docstring we replace reasons and origin_ref. Note that if we loose them, isn't it a bit sad to log them then forget them ? Then, shouldn't we inherit on mail.thread and log a consolidated message with all references ?

As it is a bit outside the scope of this task, maybe deserves its own followup task and discuss with Frgi :)